python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,480 @@
+                                  "regularization currently.")
+            else:
+                warn("LARS is not currently GPU accelerated, using "
+                     "Scikit-learn.")
+
+                from sklearn.linear_model import LassoLarsIC, lars_path
+                if (isinstance(l1_reg, str)
+                        and l1_reg.startswith("num_features(")):
+                    r = int(l1_reg[len("num_features("):-1])
+                    nonzero_inds = lars_path(
+                        self.mask, y_hat, max_iter=r)[1]
+                elif (isinstance(l1_reg, str) and l1_reg == "bic" or
+                        l1_reg == "aic"):
+                    nonzero_inds = np.nonzero(
+                        LassoLarsIC(criterion=l1_reg).fit(self.mask,
+                                                          y_hat).coef_)[0]
+
+        # weighted linear regression
+        # todo: see wheter change to use cuML linear regression with weights
+        # todo: small optimizations
+        y_hat = y_hat - self.expected_value
+
+        y_hat = y_hat - self.mask[:, nonzero_inds[-1]] * (
+            self.link_fn(self.fx) - self.link_fn(self.expected_value)
+        )
+
+        etmp = cp.transpose(
+            cp.transpose(
+                self.mask[:,
+                          nonzero_inds[:-1]]) - self.mask[:, nonzero_inds[-1]])
+
+        tmp = cp.dot(cp.dot(etmp.T, cp.diag(self.weights)), etmp)
+
+        # cupy linalg solve requires tmp to be square and full rank,
+        # so we would need the pseudo inverse anyways
+        try:
+            tmp = cp.linalg.inv(tmp)
+        except cp.linalg.LinAlgError:
+            tmp = cp.linalg.pinv(tmp)
+        return cp.dot(tmp,
+                      cp.dot(cp.dot(etmp.T, cp.diag(self.weights)), y_hat))
+
+    def shap_values(self, X, l1_reg='auto'):
+        """
+        Legacy interface to estimate the SHAP values for a set of samples.
+
+        Parameters
+        ----------
+        X : Dense matrix containing floats or doubles.
+            Acceptable formats: CUDA array interface compliant objects like
+            CuPy, cuDF DataFrame/Series, NumPy ndarray and Pandas
+            DataFrame/Series.
+        l1_reg : str (default: 'auto')
+            The l1 regularization to use for feature selection.
+
+        Returns
+        -------
+        array or list
+
+        """
+        return self.explain(X, l1_reg)
+
+    def __call__(self, X, l1_reg='auto'):
+        warn("SHAP's Explanation object is still experimental, the main API "
+             "currently is 'explainer.shap_values'.")
+        res = self.explain(X, l1_reg)
+        out = Explanation(
+            values=res,
+            base_values=self.expected_value,
+            base_values=self.expected_value,
+            data=self.background,
+            feature_names=self.feature_names,
+        )
+        return out
+
+
+@lru_cache(maxsize=None)
+def binomCoef(n, k):
+    res = 1
+    if(k > n - k):
```

Comment:
maybe this whole function could be cython / cdef / typed for speed? Or maybe the speed is totally irrelevant